File: packages/core/src/protocol.ts
// Botiva Wire Protocol v1 ("botiva/1") — the transport-agnostic frame format.
//
// Every transport (WebSocket, Socket.IO, SSE, SignalR, ...) carries these exact
// JSON frames, so the client side needs exactly ONE connector regardless of
// transport, and server implementations in other languages (Go, .NET, Python)
// only have to reproduce this mapping. See PROTOCOL.md for the full spec.
//
// Frames are either:
//   persistent — appended to conversation history with a monotonic `seq`
//                (watermark) and replayed on reconnect: text, tool_call, genui
//   transient  — delivery-only, never replayed: welcome, run, busy notices

import type { GenUIChunk, MessageAction, AgentEvent, ToolCall } from "./events.js";

export const PROTOCOL_VERSION = "botiva/1";

// ── frames ───────────────────────────────────────────────────────────────────

/** Client → server handshake (also expressible as query params / socket.io auth). */
export interface HelloFrame {
    type: "hello";
    userId?: string;
    conversationId?: string;
    /** Highest `seq` the client has already seen; server replays newer frames. */
    watermark?: number;
    meta?: Record<string, unknown>;
}

/** Server → client, first frame after connect. Transient. */
export interface WelcomeFrame {
    type: "welcome";
    data: {
        protocol: string;
        conversationId: string;
        userId: string;
        connectionId: string;
        /** Current end of history — pass this back as `watermark` when reconnecting. */
        watermark: number;
    };
}

/** Chat bubble (both directions). Persistent. */
export interface TextFrame {
    type: "text";
    id: string;
    seq?: number;
    from: "bot" | "user";
    data: { text: string };
    /** Action chips (HITL questions, suggestions). */
    actions?: MessageAction[];
    timestamp: number;
}

/** Tool activity upsert (match by data.id). Persistent. */
export interface ToolCallFrame {
    type: "tool_call";
    seq?: number;
    data: ToolCall;
}

/** Generative UI chunk. Persistent. */
export interface GenUIFrame {
    type: "genui";
    seq?: number;
    streamId: string;
    chunk: GenUIChunk;
    done: boolean;
}

/** Run/typing status. Transient. */
export interface RunFrame {
    type: "run";
    data: { status: "started" | "finished" };
}

export type Frame = HelloFrame | WelcomeFrame | TextFrame | ToolCallFrame | GenUIFrame | RunFrame;

/** Frame types that get a `seq` and are replayed after reconnect. */
export const PERSISTENT_FRAME_TYPES = ["text", "tool_call", "genui"] as const;

// ── inbound parsing ──────────────────────────────────────────────────────────

export interface IncomingMessage {
    text: string;
    id?: string;
    meta?: Record<string, unknown>;
}

export type Inbound =
    | { kind: "hello"; hello: HelloFrame }
    | { kind: "message"; message: IncomingMessage };

/**
 * Parse anything a transport hands us: a JSON string, a parsed frame object,
 * or plain text. Returns null for empty/unrecognized input.
 */
export function parseIncoming(raw: unknown): Inbound | null {
    let value: unknown = raw;
    if (typeof raw === "string") {
        try {
            value = JSON.parse(raw);
        } catch {
            const text = raw.trim();
            return text ? { kind: "message", message: { text } } : null;
        }
    }
    if (value === null || typeof value !== "object") return null;
    const frame = value as Record<string, unknown>;
    if (frame.type === "hello") return { kind: "hello", hello: frame as unknown as HelloFrame };

    const data = frame.data as Record<string, unknown> | undefined;
    const text = String(data?.text ?? frame.text ?? "").trim();
    if (!text) return null;
    return {
        kind: "message",
        message: {
            text,
            ...(typeof frame.id === "string" ? { id: frame.id } : {}),
            ...(frame.meta && typeof frame.meta === "object"
                ? { meta: frame.meta as Record<string, unknown> }
                : {}),
        },
    };
}

// ── event → frame mapping ────────────────────────────────────────────────────

export interface FrameMapping {
    frame: Frame;
    persistent: boolean;
}

/**
 * The canonical AgentEvent → wire frame mapping. Pure function — this is the
 * part every language implementation must reproduce byte-compatibly.
 */
export function eventToFrames(ev: AgentEvent, newId: (prefix: string) => string): FrameMapping[] {
    const now = Date.now();
    const text = (t: string, extra: Partial<TextFrame> = {}): TextFrame => ({
        type: "text",
        id: newId("msg"),
        from: "bot",
        data: { text: t },
        timestamp: now,
        ...extra,
    });

    switch (ev.type) {
        case "message":
            return [{ frame: text(ev.text, ev.actions ? { actions: ev.actions } : {}), persistent: true }];
        case "tool_call":
            return [{ frame: { type: "tool_call", data: ev.toolCall }, persistent: true }];
        case "genui":
            return [
                {
                    frame: {
                        type: "genui",
                        streamId: ev.streamId ?? newId("stream"),
                        chunk: ev.chunk,
                        done: ev.done === true,
                    },
                    persistent: true,
                },
            ];
        case "interrupt": {
            const p = (ev.payload ?? {}) as Record<string, unknown> | string;
            const question =
                typeof p === "string"
                    ? p
                    : String(p.question ?? p.message ?? "Your confirmation is needed to continue.");
            const rawOptions =
                typeof p === "object" && Array.isArray(p.options) ? p.options : ["Approve", "Cancel"];
            const actions: MessageAction[] = rawOptions.map((o: unknown) =>
                typeof o === "string" ? { label: o } : (o as MessageAction),
            );
            return [{ frame: text(question, { actions }), persistent: true }];
        }
        case "busy":
            return [
                {
                    frame: text("⏳ Still working on the previous message — one moment."),
                    persistent: false,
                },
            ];
        case "run_started":
            return [{ frame: { type: "run", data: { status: "started" } }, persistent: false }];
        case "run_finished":
            return [{ frame: { type: "run", data: { status: "finished" } }, persistent: false }];
        case "run_error":
            return [
                { frame: text(`⚠️ ${ev.error}`), persistent: true },
                { frame: { type: "run", data: { status: "finished" } }, persistent: false },
            ];
        default:
            return [];
    }
}

File: packages/core/python/botiva/protocol.py
"""Botiva Wire Protocol v1 — frame parsing + the canonical event→frame mapping.

Must stay byte-compatible with @botiva/core (PROTOCOL.md §3–4). Frames are
plain dicts; persistent frames get a monotonic ``seq`` from the HistoryStore.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Callable

from .events import AgentEvent, now_ms

PROTOCOL_VERSION = "botiva/1"
PERSISTENT_FRAME_TYPES = ("text", "tool_call", "genui")

Frame = dict[str, Any]


@dataclass
class IncomingMessage:
    text: str
    id: str | None = None
    meta: dict[str, Any] | None = None


@dataclass
class Hello:
    user_id: str | None = None
    conversation_id: str | None = None
    watermark: int | None = None
    token: str | None = None  # auth credential (§2.1), when the server authenticates
    meta: dict[str, Any] | None = None


@dataclass
class Inbound:
    hello: Hello | None = None
    message: IncomingMessage | None = None


def parse_incoming(raw: Any) -> Inbound | None:
    """Accepts a JSON string, a parsed frame dict, bytes or plain text."""
    value: Any = raw
    if isinstance(raw, (bytes, bytearray)):
        value = raw.decode("utf-8", errors="replace")
    if isinstance(value, str):
        try:
            value = json.loads(value)
        except json.JSONDecodeError:
            text = value.strip()
            return Inbound(message=IncomingMessage(text=text)) if text else None
    if not isinstance(value, dict):
        return None
    if value.get("type") == "hello":
        watermark = value.get("watermark")
        return Inbound(hello=Hello(
            user_id=value.get("userId"),
            conversation_id=value.get("conversationId"),
            watermark=int(watermark) if watermark is not None else None,
            token=value.get("token"),
            meta=value.get("meta"),
        ))
    data = value.get("data") if isinstance(value.get("data"), dict) else {}
    text = str(data.get("text") or value.get("text") or "").strip()
    if not text:
        return None
    return Inbound(message=IncomingMessage(
        text=text,
        id=value.get("id") if isinstance(value.get("id"), str) else None,
        meta=value.get("meta") if isinstance(value.get("meta"), dict) else None,
    ))


def error_frame(code: str, message: str) -> Frame:
    """Build a transient ``error`` frame (auth rejection, protocol error)."""
    return {"type": "error", "data": {"code": code, "message": message}}


@dataclass
class FrameMapping:
    frame: Frame
    persistent: bool


def event_to_frames(ev: AgentEvent, new_id: Callable[[str], str]) -> list[FrameMapping]:
    """The canonical AgentEvent → wire frame mapping (PROTOCOL.md §4)."""
    now = now_ms()

    def text_frame(text: str, **extra: Any) -> Frame:
        return {
            "type": "text",
            "id": new_id("msg"),
            "from": "bot",
            "data": {"text": text},
            "timestamp": now,
            **extra,
        }

    kind = ev.get("type")
    if kind == "message":
        extra = {"actions": ev["actions"]} if ev.get("actions") else {}
        return [FrameMapping(text_frame(ev.get("text", ""), **extra), True)]
    if kind == "tool_call":
        return [FrameMapping({"type": "tool_call", "data": ev.get("toolCall")}, True)]
    if kind == "genui":
        return [FrameMapping({
            "type": "genui",
            "streamId": ev.get("streamId") or new_id("stream"),
            "chunk": ev.get("chunk"),
            "done": ev.get("done") is True,
        }, True)]
    if kind == "interrupt":
        payload = ev.get("payload") or {}
        if isinstance(payload, str):
            question, options = payload, ["Approve", "Cancel"]
        else:
            question = str(payload.get("question") or payload.get("message")
                           or "Your confirmation is needed to continue.")
            options = payload.get("options") or ["Approve", "Cancel"]
        actions = [{"label": o} if isinstance(o, str) else o for o in options]
        return [FrameMapping(text_frame(question, actions=actions), True)]
    if kind == "busy":
        return [FrameMapping(text_frame("⏳ Still working on the previous message — one moment."), False)]
    if kind == "run_started":
        return [FrameMapping({"type": "run", "data": {"status": "started"}}, False)]
    if kind == "run_finished":
        return [FrameMapping({"type": "run", "data": {"status": "finished"}}, False)]
    if kind == "run_error":
        return [
            FrameMapping(text_frame(f"⚠️ {ev.get('error')}"), True),
            FrameMapping({"type": "run", "data": {"status": "finished"}}, False),
        ]
    return []
